### PR TITLE
OADP-612: Check VSB status conditions for failure

### PR DESCRIPTION
**Pull request by eemcmullan**
Related: https://github.com/openshift/velero/pull/192

### CODE REVIEW

File: internal/util/util.go [hunk 1]
```diff
@@ -491,3 +491,56 @@ func VSBHasVSBackupName(backup *velerov1api.Backup, snapCont *snapshotv1api.Volu
...
+	doesVSBHaveConditions, err := GetVSBWithConditions(vsb.Namespace, vsb.Name, log)
+	if err != nil {
+		return err
+	}
+
+	if !doesVSBHaveConditions {
+		return errors.Errorf("err getting conditions for volumesnapshotbackup %s ", vsb.Name)
+	}
+
+	for _, condition := range vsb.Status.Conditions {
```
Comment (shubham-pampattiwar) [inline internal/util/util.go:540]: Should you not check for condition type here ?

File: internal/util/util.go [hunk 2]
```diff
@@ -491,3 +491,56 @@ func VSBHasVSBackupName(backup *velerov1api.Backup, snapCont *snapshotv1api.Volu
 
 	return true
 }
+
+func GetVSBWithConditions(volumeSnapshotbackupNS string, volumeSnapshotName string, log logrus.FieldLogger) (bool, error) {
```
Comment (shubham-pampattiwar) [inline internal/util/util.go:495]: Do we really need to poll here ? Should we not directly fetch VSB ?

File: internal/backup/volumesnapshotbackup_action.go [hunk 3]
```diff
@@ -34,8 +34,14 @@ func (p *VolumeSnapshotBackupBackupItemAction) Execute(item runtime.Unstructured
 		return nil, nil, errors.WithStack(err)
 	}
 	p.Log.Infof("Converted Item to VSB: %v", vsb)
-	vsbNew, err := util.GetVolumeSnapshotbackupWithStatusData(vsb.Namespace, vsb.Name, p.Log)
 
+	// check if VSB has failed before continuing
+	err := util.CheckVSBFailed(backup, &vsb, p.Log)
```
Comment (shubham-pampattiwar) [inline internal/backup/volumesnapshotbackup_action.go:39]: We are waiting for status data here as well right ? Similar to `GetVolumeSnapshotBackupWithStatusData`

File: internal/util/util.go (hunk 2 above)
Comment (shubham-pampattiwar) [reply internal/util/util.go:495]: This seems like code duplication, lets modify the GetVolumeSnapshotBackupWithStatusData function itself,
like once you get vsb with status data, check for condition using `checkVSBFailed` func and fail the backup there itself.

**Review by shubham-pampattiwar — CHANGES_REQUESTED** [review]

File: internal/backup/volumesnapshotbackup_action.go (hunk 3 above)
Comment (eemcmullan) [reply internal/backup/volumesnapshotbackup_action.go:39]: I'll modify this then

**Review by weshayutin — COMMENTED** [review]
Looks good to me, but too much of a newb to approve w/o testing locally.

File: internal/util/util.go [hunk 4]
```diff
@@ -326,6 +326,18 @@ func GetVolumeSnapshotbackupWithStatusData(volumeSnapshotbackupNS string, volume
...
 		}
 
+		if len(vsb.Status.Conditions) == 0 {
+			log.Infof("Waiting for volumesnapshotbackup %s to have conditions. Retrying in %ds", vsb.Name, interval/time.Second)
+			return false, nil
+		}
+
+		// check for status failure first
+		for _, condition := range vsb.Status.Conditions {
+			if condition.Status == metav1.ConditionFalse {
```
Comment (shubham-pampattiwar) [inline internal/util/util.go:336]: @eemcmullan Should we check for the condition type as well ?

Comment (eemcmullan) [reply internal/util/util.go:336]: @shubham-pampattiwar what about condition reason? It looks like condition type would be the same for failed and successful VSB: https://github.com/migtools/volume-snapshot-mover/blob/master/controllers/volumesnapshotbackup_controller.go#L126 

https://github.com/migtools/volume-snapshot-mover/blob/master/controllers/volumesnapshotbackup_controller.go#L137

Comment (shubham-pampattiwar) [reply internal/util/util.go:336]: +1 on the reason check along with type.

File: internal/util/util.go [hunk 5]
```diff
@@ -52,6 +52,8 @@ const (
 	//TODO: use annotation from velero https://github.com/vmware-tanzu/velero/pull/2283
 	resticPodAnnotation = "backup.velero.io/backup-volumes"
 )
+const ReconciledReasonError = "Error"
```
Comment (shubham-pampattiwar) [inline internal/util/util.go:55]: NIT: Combine all the const types together

**Review by shubham-pampattiwar — APPROVED** [review]

**Review by kaovilai — APPROVED** [review]

**Pull request merged**